### ISSUE TITLE
Inefficient x64 codegen for all_true

### ISSUE THREAD

**Issue by abrown** [WebAssembly/simd#189]
[`all_true`](https://github.com/WebAssembly/simd/blob/master/proposals/simd/SIMD.md#all-lanes-true) checks if all lanes are (unsigned) greater than 0. This requires 4 instructions in [cranelift](https://github.com/bytecodealliance/cranelift/blob/48029b4a16264672ce24afbee1050b37e1e68020/cranelift-codegen/meta/src/isa/x86/legalize.rs#L471-L501) and 6 in [v8](https://github.com/v8/v8/blob/19be4913881bb02c5d9b4f1c7547ee2d1273120b/src/compiler/backend/x64/code-generator-x64.cc#L590-L602). Perhaps there is a more granular way to reduce lanes (see [movemask](https://github.com/WebAssembly/simd/issues/131)) and avoid this inefficiency? 

(Along these lines, `any_true` is 4 instructions in [v8](https://github.com/v8/v8/blob/19be4913881bb02c5d9b4f1c7547ee2d1273120b/src/compiler/backend/x64/code-generator-x64.cc#L3909-L3922) and could be 2 as in [cranelift](https://github.com/bytecodealliance/cranelift/blob/48029b4a16264672ce24afbee1050b37e1e68020/cranelift-codegen/meta/src/isa/x86/legalize.rs#L461-L468) with the use of `SETcc`).

**Comment by ngzhian** [comment]
setcc will only set the low byte of the register, how does cranelift deal with this? Is the dst register zero-ed first?

**Comment by abrown** [comment]
Cranelift's SSA form has types for each value; if you try to use an `I8` value somewhere as an `I64` then it will extend it appropriately. (That is my hazy recollection from when I implemented it a long, long time ago).

**Comment by abrown** [comment]
But looking at the V8 implementation: why not XOR the tmp register as is done currently and use setcc?

**Comment by ngzhian** [comment]
> it will extend it appropriately.
Got it, thanks! So there's one more instruction for extension. In v8 I used xor, which I guess is the same in terms of instruction counts.

I think you're looking at the older version, the latest one uses setcc https://github.com/v8/v8/blob/master/src/compiler/backend/x64/code-generator-x64.cc#L604-L613

I couldn't do the same on IA-32, setcc there requires a byte register, and I don't think we can specify that we want a byte register in our register allocation, can only check it. I looked briefly at cranelift, it looks like it only supports x86-64, so this is not a concern there, right?

**Comment by abrown** [comment]
I wasn't too focused on IA-32. Thanks for the link to the latest; I guess opening some of these issues had some effect after all! I have been waiting for a document to put these implementation notes in... @ngzhian, @tlively, what do you think about adding them to the SIMD document itself but as collapsible sections, like:

<details><summary>Implementation notes:</summary>
<h5>x86/x86-64 processors with AVX instruction sets</h5>
<p>TODO</p>
<h5>x86/x86-64 processors with SSE4.1 instruction sets</h5>
<p>TODO</p>
<h5>ARM64 processors</h5>
<p>TODO</p>
</details>

That way both the implementors and the users can quickly see lowerings. I would assume users might actually want a hint as to what these instructions will compile down to, even if different runtimes actually emit different things.

**Comment by ngzhian** [comment]
>  I guess opening some of these issues had some effect after all!

Definitely, thank you for doing so. A lot of the issues don't have workarounds, that's how the instructions are spec-ed and how we need to implement them.

I don't remember what these "implementation notes" are for? Is it guidance for engine implementors to make sure their implementations are fast?

**Comment by abrown** [comment]
Yeah, that and to alert users that the semantics of certain instructions may have unintended codegen; maybe you use a Wasm SIMD instruction in your performance-critical code and then you realize that it doesn't perform as well as you thought on certain architectures. I think it would be good to be upfront about that rather than having to dig through v8 source code, e.g., as I have had to do.

**Comment by ngzhian** [comment]
>  I think it would be good to be upfront about that rather than having to dig through v8 source code, e.g., as I have had to do.

@zeux has this https://github.com/zeux/wasm-simd/ which targets part of it. I think it's an appropriate place for such notes, provided we keep it up to date :) I don't think we should do that work *right now*, since the instruction set is still in flux. After we finalize (hard freeze), we can probably invest more time into it.

**Comment by abrown** [comment]
Yeah, in previous discussions I've mentioned that I wanted that information to be more visible and official. I'm not opposed to waiting longer, though (less work right now 😄).

**Comment by ngzhian** [comment]
> Yeah, in previous discussions I've mentioned that I wanted that information to be more visible and official.

Ah okay, sorry I've lost track of this. Perhaps during that discussion it was brought up that the spec might not be the right place for this sort of information? Maybe it can be in the Appendix, which is the only place containing [implementation-related information](https://webassembly.github.io/spec/core/appendix/implementation.html).

> I'm not opposed to waiting longer

Same, let's leave this as it is for now :)

**Comment by abrown** [comment]
Yeah, that appendix seems like the right type of thing--in the same general place as the specification but not cluttering the semantics.

**Comment by Maratyszcza** [comment]
> why not XOR the tmp register as is done currently and use setcc?

Partial register updates are very expensive on Intel. It would be more efficient to do `setcc r8` + movzx r32, r8`, even though it increase latency of the dependency chain.
